### PR TITLE
docs(web-components): implementation notes for Web Components

### PR DESCRIPTION
**Pull request by asudoh**
### Related Ticket(s)

Refs #4916.

### Description

Adds documentation about the implementation details of `@carbon/ibmdotcom-web-components` so new developers can easily jump into the codebase. More to be added as separate commits.

### Changelog

**Changed**

- Initial documentation about the implementation details of `@carbon/ibmdotcom-web-components`.

### CODE REVIEW

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+| [`<dds-left-nav>`](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/masthead/left-nav.ts)                                 | [`<bx-side-nav>`](https://github.com/carbon-design-system/carbon-web-components/blob/master/src/components/ui-shell/side-nav.ts)     |
+| [`<dds-language-selector-desktop>`](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/footer/language-selector-desktop.ts) | [`<bx-combo-box>`](https://github.com/carbon-design-system/carbon-web-components/blob/master/src/components/combo-box/combo-box.ts)  |
+| [`<dds-language-selector-mobile>`](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/footer/language-selector-mobile.ts)   | [`<bx-select>`](https://github.com/carbon-design-system/carbon-web-components/blob/master/src/components/select/select.ts)           |
+
+## Component CSS
+
+### Usage of Sass
+
+Most components has the following to define CSS:
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:42]:
```suggestion
Most components have the following to define CSS:
```

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+class DDSSomeComponent extends LitElement {
+  ...
+
+  static styles = styles;
+}
+```
+
+There are several things behind the scene of that.
+
+- In Storybook environment, in addition to `sass-loader`/`postcss-loader`/`style-loader` chains, we use a [custom WebPack loader](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/tools/css-result-loader.js) that generates `lit-html`'s [CSS tagged template](https://lit-element.polymer-project.org/guide/styles#add-styles) from the generated CSS.
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:57]:
```suggestion
- In Storybook environments, in addition to `sass-loader`/`postcss-loader`/`style-loader` chains, we use a [custom WebPack loader](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/tools/css-result-loader.js) that generates `lit-html`'s [CSS tagged template](https://lit-element.polymer-project.org/guide/styles#add-styles) from the generated CSS.
```

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+- For production build, we [generate `.css.js` file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/gulp-tasks/build.js#L79-L88) that exports `lit-html`'s [CSS tagged template](https://lit-element.polymer-project.org/guide/styles#add-styles) with the CSS compiled from the Sass code. The [`import` path is changed to refer to such generated `.css.js` file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/tools/babel-plugin-resource-js-paths.js#L24-L26).
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:58]:
```suggestion
- For production builds, we [generate `.css.js` file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/gulp-tasks/build.js#L79-L88) that exports `lit-html`'s [CSS tagged template](https://lit-element.polymer-project.org/guide/styles#add-styles) with the CSS compiled from the Sass code. The [`import` path is changed to refer to such generated `.css.js` file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.13.0/packages/web-components/tools/babel-plugin-resource-js-paths.js#L24-L26).
```

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+Above custom toolchain allows us to use Sass for our components, as well as to put Sass/CSS code in a separate location from the component code.
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:60]:
```suggestion
The above custom toolchain allows us to use Sass for our components, as well as to put Sass/CSS code in a separate location from the component code.
```

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+### Styles in Shadow DOM
+
+One of the greatest features in Web Components is that CSS is isolated within components, thanks to Shadow DOM. Component CSS doesn't affect application CSS, or vice versa. This allows user to do gradual migration from `v18` design assets from Carbon for IBM.com design assets, for example.
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:64]:
```suggestion
One of the greatest features in Web Components is that CSS is isolated within components, thanks to Shadow DOM. Component CSS doesn't affect application CSS, or vice versa. This allows users to gradually migrate from `v18` design assets to Carbon for IBM.com design assets, for example.
```

File: packages/web-components/IMPLEMENTATION_NOTES.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+Style isolation within component causes challenges in implementing component styles, in some cases. One of the biggest challenge is a component that requires different style if it's put in another component's child.
+
+Style defined in the shadow DOM can define the style of the direct child with [`::slotted()` pseudo element](https://developer.mozilla.org/en-US/docs/Web/CSS/::slotted), but [cannot define the style of the descendants](https://github.com/WICG/webcomponents/issues/745).
+
+There are a couple approaches `@carbon/ibmdotcom-web-components` chooses from depending on what best suits the use case:
```
Comment (kennylam) [inline packages/web-components/IMPLEMENTATION_NOTES.md:70]:
```suggestion
There are several approaches `@carbon/ibmdotcom-web-components` to choose from depending on what best suits the use case:
```

**Review by kennylam — APPROVED** [review]
Looks good thanks @asudoh! Just a few typos.

**Comment by asudoh** [comment]
Thank you for reviewing and comments @kennylam!

**Pull request merged**